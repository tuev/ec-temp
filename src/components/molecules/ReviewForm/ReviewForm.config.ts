import { IReviewFormValidator } from './ReviewForm.types'

export const ReviewFormConfig: IReviewFormValidator = {
  nickname: {
    required: true,
    validator: {
      func: (value: string): boolean => /^[a-zA-Z]+$/.test(value),
      error: 'Invalid first name format.',
    },
  },
  reviewText: {
    required: true,
    validator: {
      func: (value: string): boolean => value.length > 50,
      error: 'Review must be at least 50 characters.',
    },
  },
  rating: {
    required: true,
    validator: {
      func: (value: string): boolean => Number(value) > 0,
      error: 'Rating must be greater 0.',
    },
  },
}
