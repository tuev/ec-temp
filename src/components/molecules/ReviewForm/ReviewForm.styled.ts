/**
 *
 * Styled for ReviewForm
 *
 */

import styled from 'styled-components'
