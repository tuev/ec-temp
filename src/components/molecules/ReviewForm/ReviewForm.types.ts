import { ChangeEvent } from 'react'

export interface IReviewField {
  value?: string | number
  error?: string
  dirty: boolean
}

export interface IReviewForm {
  nickname: IReviewField
  reviewText: IReviewField
  rating: IReviewField
  [key: string]: IReviewField
}
export interface IReviewValidatorRule {
  func(value?: string | number): boolean
  error: string
  [key: number]: string | boolean
}
export interface IReviewFieldValidator {
  required?: boolean
  validator?: IReviewValidatorRule
}

export interface IReviewFormValidator {
  nickname: IReviewFieldValidator
  reviewText: IReviewFieldValidator
  rating: IReviewFieldValidator
  [key: string]: IReviewFieldValidator
}

export type IReviewHookForm = (
  stateSchema: IReviewForm,
  stateValidatorSchema: IReviewFormValidator,
  submitFormCallback: (data: ISubmittedData) => void
) => {
  handleOnChange(event?: ChangeEvent): void
  handleOnSubmit(
    event?: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ): void | undefined
  handleOnRating(event: React.FormEvent<HTMLSpanElement>): void
  state: IReviewForm
  disable: boolean
}
export interface IReviewAction {
  type: string
  payload: {
    name: string
    data?: string | number
    error?: string
  }
}

export type ReviewFormReducer = (
  initState: IReviewForm,
  action: IReviewAction
) => IReviewForm

export type ValidationFormFields = (
  name: string,
  value?: string | number
) => string | undefined

export interface ISubmittedData {
  nickname: string
  reviewText: string
  rating: number
  [key: string]: string | number | undefined
}
