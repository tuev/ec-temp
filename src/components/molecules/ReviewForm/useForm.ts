import { useState, useEffect, useCallback } from 'react'
import {
  IReviewFormValidator,
  IReviewValidatorRule,
  IReviewForm,
  IReviewFilterData,
  IReviewHookForm,
  IReviewDirty,
} from './ReviewForm.types'

const VALUE = 'value'
const ERROR = 'error'
const REQUIRED_FIELD_ERROR = 'This is required field'

/**
 * Determines a value if it's an object
 *
 * @param {object} value
 */
function isObject(value: IReviewValidatorRule | undefined): boolean {
  return typeof value === 'object' && value !== null
}

/**
 *
 * @param {string} value
 * @param {boolean} isRequired
 */
function isRequired(value: unknown, isRequired?: boolean): string {
  if (!value && isRequired) return REQUIRED_FIELD_ERROR
  return ''
}

function getPropValues(
  stateSchema: IReviewForm,
  prop: string
): IReviewFilterData {
  return Object.keys(stateSchema).reduce((accumulator, curr) => {
    accumulator[curr] = !prop ? '' : stateSchema[curr][prop]
    return accumulator
  }, {} as IReviewFilterData)
}

function getDirtyState(stateSchema: IReviewForm): IReviewDirty {
  return Object.keys(stateSchema).reduce((accumulator, curr) => {
    accumulator[curr] = false
    return accumulator
  }, {} as IReviewDirty)
}

/**
 * Custom hooks to validate your Form...
 *
 * @param {object} stateSchema model you stateSchema.
 * @param {object} stateValidatorSchema model your validation.
 * @param {function} submitFormCallback function to be execute during form submission.
 */

const useForm = (
  stateSchema: IReviewForm,
  stateValidatorSchema: IReviewFormValidator,
  submitFormCallback: Function
): IReviewHookForm => {
  const [state, setStateSchema] = useState(stateSchema)

  const [values, setValues] = useState(getPropValues(state, VALUE))
  const [errors, setErrors] = useState(getPropValues(state, ERROR))
  const [dirty, setDirty] = useState(getDirtyState(state))

  const [disable, setDisable] = useState(true)
  const [isDirty, setIsDirty] = useState(false)

  const validateFormFields = useCallback(
    (name: string, value: string | number) => {
      const validator: IReviewFormValidator = stateValidatorSchema
      if (!validator[name]) return ''

      const field = validator[name]

      let error: string | number = ''
      error = isRequired(value, field.required)

      const fieldValidator: IReviewValidatorRule | undefined =
        field['validator']
      if (fieldValidator && isObject(fieldValidator) && error === '') {
        // Test the function callback if the value is meet the criteria
        const validateInput: Function = fieldValidator['func']
        if (!validateInput(value)) {
          error = fieldValidator['error']
        }
      }

      return error
    },
    [stateValidatorSchema]
  )

  const setInitialErrorState = useCallback(() => {
    Object.keys(errors).map(name => {
      setErrors(prevState => ({
        ...prevState,
        [name]: validateFormFields(name, values[name]),
      }))
    })
  }, [errors, values, validateFormFields])

  useEffect(() => {
    setStateSchema(stateSchema)
    setDisable(true)
    setInitialErrorState()
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const validateErrorState = useCallback(() => {
    return Object.values(errors).some(value => value)
  }, [errors])

  // For every changed in our state this will be fired
  // To be able to disable the button
  useEffect(() => {
    if (isDirty) {
      setDisable(validateErrorState())
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isDirty, errors])

  const commonChange = useCallback(
    (name: string, value: string | number): void => {
      const error = validateFormFields(name, value)

      setValues(prevState => ({ ...prevState, [name]: value }))
      setErrors(prevState => ({ ...prevState, [name]: error }))
      setDirty(prevState => ({ ...prevState, [name]: true }))
    },
    [validateFormFields]
  )

  const handleOnChange = useCallback(
    event => {
      setIsDirty(true)

      const name = event.target.name
      const value = event.target.value

      commonChange(name, value)
    },
    [commonChange]
  )

  const handleOnRating = useCallback(
    event => {
      setIsDirty(true)
      const name = 'rating'
      const value = event.target.value

      commonChange(name, value)
    },
    [commonChange]
  )

  const handleOnSubmit = useCallback(() => {
    if (!validateErrorState()) {
      submitFormCallback(values)
    }
  }, [validateErrorState, submitFormCallback, values])

  return {
    handleOnChange,
    handleOnSubmit,
    handleOnRating,
    values,
    errors,
    disable,
    dirty,
  }
}

export default useForm
