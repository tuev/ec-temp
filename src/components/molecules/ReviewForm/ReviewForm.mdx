# Stories mdx for ReviewForm

import { Meta, Story } from '@storybook/addon-docs/blocks'
import * as stories from './ReviewForm.stories.tsx'
import ReviewForm from './ReviewForm.tsx'

<Story name="basic" />