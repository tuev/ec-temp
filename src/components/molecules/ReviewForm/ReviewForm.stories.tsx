/**
 *
 *  Stories for ReviewForm
 *
 */

import React from 'react'
import { action } from '@storybook/addon-actions'
import ReviewForm from './ReviewForm'

export default {
  component: ReviewForm,
  includeStories: ['basic'],
  title: 'ReviewForm',
}

export const basic = () => (
  <div>
    <ReviewForm
      id="123"
      authorId="01"
      authorName="dai"
      authorRating={2}
      authorOpinion="opinion"
    />
  </div>
)
