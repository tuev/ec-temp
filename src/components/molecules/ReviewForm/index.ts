/**
 *
 * Index for ReviewForm
 *
 */

import ReviewForm from './ReviewForm'

export default ReviewForm
