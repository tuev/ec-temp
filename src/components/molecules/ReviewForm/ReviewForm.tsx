/**
 *
 * ReviewForm
 *
 */

import React, { FC } from 'react'
import {
  IReviewForm,
  IReviewFormValidator,
  ISubmittedData,
} from './ReviewForm.types'
import { Grid } from '@material-ui/core'
import { Input, TextArea, Rating, Button } from '../../atoms'
import { Title, Text, Error } from './ReviewForm.styled'
import useForm from './useForm'

const ReviewForm: FC = () => {
  const stateReview: IReviewForm = {
    nickname: { value: '', dirty: false },
    reviewText: { value: '', dirty: false },
    rating: { value: 0, dirty: false },
  }

  const stateValidatorSchema: IReviewFormValidator = {
    nickname: {
      required: true,
      validator: {
        func: (value: string): boolean => /^[a-zA-Z]+$/.test(value),
        error: 'Invalid first name format.',
      },
    },
    reviewText: {
      required: true,
      validator: {
        func: (value: string): boolean => value.length > 50,
        error: 'Review must be at least 50 characters.',
      },
    },
    rating: {
      required: true,
      validator: {
        func: (value: string): boolean => Number(value) > 0,
        error: 'Rating must be greater 0.',
      },
    },
  }

  const onSubmitForm = (data: ISubmittedData): void => {
    const value = data
    console.log(value)
  }

  const {
    state,
    handleOnChange,
    handleOnRating,
    handleOnSubmit,
    disable,
  } = useForm(stateReview, stateValidatorSchema, onSubmitForm)

  const { nickname, reviewText, rating } = state
  return (
    <div>
      <Title customvariant="header2" mb={4}>
        Add a review
      </Title>
      <Grid container={true} spacing={3}>
        <Grid container={true} item={true} alignItems="center">
          <Grid item={true} xs={12} md={3}>
            <Text customvariant="body1">Choose a nickname</Text>
          </Grid>
          <Grid item={true} xs={12} md={9}>
            <Input
              name="nickname"
              value={nickname.value}
              onChange={handleOnChange}
              error={Boolean(nickname.error) && nickname.dirty}
            />
            {Boolean(nickname.error) && nickname.dirty && (
              <Error customvariant="body2">{nickname.error}</Error>
            )}
          </Grid>
        </Grid>

        <Grid container={true} item={true}>
          <Grid item={true} xs={12} md={3}>
            <Text pt={10} customvariant="body1">
              Your review
            </Text>
          </Grid>
          <Grid item={true} xs={12} md={9}>
            <TextArea
              fullWidth={true}
              name="reviewText"
              onChange={handleOnChange}
              value={reviewText.value}
              error={Boolean(reviewText.error) && reviewText.dirty}
            />
            {Boolean(reviewText.error) && reviewText.dirty && (
              <Error customvariant="body2">{reviewText.error}</Error>
            )}
          </Grid>
        </Grid>

        <Grid container={true} item={true} alignItems="center">
          <Grid item={true} xs={12} md={3}>
            <Text customvariant="body1">Overall rating</Text>
          </Grid>
          <Grid item={true} xs={12} md={9}>
            <Rating value={Number(rating.value)} onInput={handleOnRating} />
          </Grid>
        </Grid>

        <Grid container={true} item={true} justify="flex-end">
          <Grid item={true} xs={12} md={9}>
            <Button
              size="medium"
              color="primary"
              variant="contained"
              disabled={disable}
              onClick={handleOnSubmit}
            >
              Submit
            </Button>
          </Grid>
        </Grid>
      </Grid>
    </div>
  )
}

export default ReviewForm
