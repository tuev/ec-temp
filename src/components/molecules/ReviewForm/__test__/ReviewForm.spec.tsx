/**
 *
 * Tests for ReviewForm
 *
 */

import React from 'react'

import { render } from '@testing-library/react'
import ReviewForm from '..'

describe('App ReviewForm', () => {
  it('Test get default ReviewForm no params ', () => {
    // test snapshot
    const wrapper = render(<ReviewForm />)
    expect(wrapper).toMatchSnapshot()
  })
})
