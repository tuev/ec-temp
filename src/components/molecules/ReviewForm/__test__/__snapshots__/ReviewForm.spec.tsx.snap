// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App ReviewForm Test get default ReviewForm no params  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c6 {
  min-width: 163px !important;
  min-height: 48px !important;
  font-size: 14px;
  border-radius: 32px;
}

.c0 {
  font-size: 24px;
  line-height: 1.5;
  font-family: 'Montserrat Alternates';
  font-weight: bold;
  color: #2E2E2E;
}

.c1 {
  font-size: 16px;
  line-height: 1.5;
  font-family: 'Open Sans';
  color: #2E2E2E;
}

.c3 {
  padding-top: 10px;
  font-size: 16px;
  line-height: 1.5;
  font-family: 'Open Sans';
  color: #2E2E2E;
}

.c4 {
  font-size: 12px;
  line-height: 2;
  font-family: 'Open Sans';
  color: #2E2E2E;
}

.c2 {
  font-weight: 600;
}

.c5 {
  color: #B4B4B4;
}

<body>
    <div>
      <div>
        <p
          class="MuiTypography-root c0 MuiTypography-body1"
          customvariant="header2"
        >
          Add a review
        </p>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3"
            >
              <p
                class="MuiTypography-root c1 c2 MuiTypography-body1"
                customvariant="body1"
              >
                Choose a nickname
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9"
            >
              <div
                class="MuiFormControl-root MuiTextField-root "
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    name="nickname"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-174 MuiOutlinedInput-notchedOutline"
                    style="padding-left: 8px;"
                  >
                    <legend
                      class="PrivateNotchedOutline-legend-175"
                      style="width: 0.01px;"
                    >
                      <span>
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3"
            >
              <p
                class="MuiTypography-root c3 c2 MuiTypography-body1"
                customvariant="body1"
                pt="10"
              >
                Your review
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                    name="reviewText"
                    rows="4"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-218 MuiOutlinedInput-notchedOutline"
                    style="padding-left: 8px;"
                  >
                    <legend
                      class="PrivateNotchedOutline-legend-219"
                      style="width: 0.01px;"
                    >
                      <span>
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <p
                class="MuiTypography-root c4 c5 MuiTypography-body1"
                customvariant="body2"
              >
                Your review must be at least 50 characters
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3"
            >
              <p
                class="MuiTypography-root c1 c2 MuiTypography-body1"
                customvariant="body1"
              >
                Overall rating
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9"
            >
              <span
                class="MuiRating-root "
              >
                <label
                  class="MuiRating-label"
                  for="app-rating-1"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconEmpty"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    1 Star
                  </span>
                </label>
                <input
                  class="MuiRating-visuallyhidden"
                  id="app-rating-1"
                  name="app-rating"
                  type="radio"
                  value="1"
                />
                <label
                  class="MuiRating-label"
                  for="app-rating-2"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconEmpty"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    2 Stars
                  </span>
                </label>
                <input
                  class="MuiRating-visuallyhidden"
                  id="app-rating-2"
                  name="app-rating"
                  type="radio"
                  value="2"
                />
                <label
                  class="MuiRating-label"
                  for="app-rating-3"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconEmpty"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    3 Stars
                  </span>
                </label>
                <input
                  class="MuiRating-visuallyhidden"
                  id="app-rating-3"
                  name="app-rating"
                  type="radio"
                  value="3"
                />
                <label
                  class="MuiRating-label"
                  for="app-rating-4"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconEmpty"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    4 Stars
                  </span>
                </label>
                <input
                  class="MuiRating-visuallyhidden"
                  id="app-rating-4"
                  name="app-rating"
                  type="radio"
                  value="4"
                />
                <label
                  class="MuiRating-label"
                  for="app-rating-5"
                >
                  <span
                    class="MuiRating-icon MuiRating-iconEmpty"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                      focusable="false"
                      role="presentation"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiRating-visuallyhidden"
                  >
                    5 Stars
                  </span>
                </label>
                <input
                  class="MuiRating-visuallyhidden"
                  id="app-rating-5"
                  name="app-rating"
                  type="radio"
                  value="5"
                />
              </span>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained c6 MuiButton-containedPrimary Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Submit
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p
        class="MuiTypography-root sc-AykKJ bXFeGJ sc-AykKK gcTcIC MuiTypography-body1"
        customvariant="header2"
      >
        Add a review
      </p>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3"
          >
            <p
              class="MuiTypography-root sc-AykKJ iexqoL sc-fzXfLO jEaXPj MuiTypography-body1"
              customvariant="body1"
            >
              Choose a nickname
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9"
          >
            <div
              class="MuiFormControl-root MuiTextField-root sc-AykKE fbCwcn"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input"
                  name="nickname"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-174 MuiOutlinedInput-notchedOutline"
                  style="padding-left: 8px;"
                >
                  <legend
                    class="PrivateNotchedOutline-legend-175"
                    style="width: 0.01px;"
                  >
                    <span>
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3"
          >
            <p
              class="MuiTypography-root sc-AykKJ kxCBjo sc-fzXfLO jEaXPj MuiTypography-body1"
              customvariant="body1"
              pt="10"
            >
              Your review
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9"
          >
            <div
              class="MuiFormControl-root MuiTextField-root sc-AykKF hjTiPI MuiFormControl-fullWidth"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
                  name="reviewText"
                  rows="4"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-218 MuiOutlinedInput-notchedOutline"
                  style="padding-left: 8px;"
                >
                  <legend
                    class="PrivateNotchedOutline-legend-219"
                    style="width: 0.01px;"
                  >
                    <span>
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <p
              class="MuiTypography-root sc-AykKJ ckspux sc-fzXfLP ePYyeJ MuiTypography-body1"
              customvariant="body2"
            >
              Your review must be at least 50 characters
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3"
          >
            <p
              class="MuiTypography-root sc-AykKJ iexqoL sc-fzXfLO jEaXPj MuiTypography-body1"
              customvariant="body1"
            >
              Overall rating
            </p>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9"
          >
            <span
              class="MuiRating-root sc-AykKH leYRC"
            >
              <label
                class="MuiRating-label"
                for="app-rating-1"
              >
                <span
                  class="MuiRating-icon MuiRating-iconEmpty"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  1 Star
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="app-rating-1"
                name="app-rating"
                type="radio"
                value="1"
              />
              <label
                class="MuiRating-label"
                for="app-rating-2"
              >
                <span
                  class="MuiRating-icon MuiRating-iconEmpty"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  2 Stars
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="app-rating-2"
                name="app-rating"
                type="radio"
                value="2"
              />
              <label
                class="MuiRating-label"
                for="app-rating-3"
              >
                <span
                  class="MuiRating-icon MuiRating-iconEmpty"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  3 Stars
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="app-rating-3"
                name="app-rating"
                type="radio"
                value="3"
              />
              <label
                class="MuiRating-label"
                for="app-rating-4"
              >
                <span
                  class="MuiRating-icon MuiRating-iconEmpty"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  4 Stars
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="app-rating-4"
                name="app-rating"
                type="radio"
                value="4"
              />
              <label
                class="MuiRating-label"
                for="app-rating-5"
              >
                <span
                  class="MuiRating-icon MuiRating-iconEmpty"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiRating-visuallyhidden"
                >
                  5 Stars
                </span>
              </label>
              <input
                class="MuiRating-visuallyhidden"
                id="app-rating-5"
                name="app-rating"
                type="radio"
                value="5"
              />
            </span>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-xs-flex-end"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-9"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-AykKI evYzjc MuiButton-containedPrimary Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Submit
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
